content/utils.go

```diff
@@ -0,0 +1,52 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package content
+
+import (
+	"errors"
+	"io"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/internal/ioutil"
+)
+
+var (
+	ErrUnexpectedEOF    = errors.New("unexpected EOF")
+	ErrTrailingData     = errors.New("trailing data")
+	ErrMismatchedDigest = errors.New("mismatched digest")
+)
+
+// ReadAll safely reads the content described by the descriptor.
+// The read content is verified against the size and the digest.
+func ReadAll(r io.Reader, desc ocispec.Descriptor) ([]byte, error) {
+	// verify while reading
+	verifier := desc.Digest.Verifier()
+	r = io.TeeReader(r, verifier)
+	buf := make([]byte, desc.Size)
+	_, err := io.ReadFull(r, buf)
+	// verify the size of the read content
+	if err != nil {
+		return nil, ErrUnexpectedEOF
```

Comment:
We should return the original error here.

Reply:
Changed accordingly.